File: internal/protocol/types/skill_publications.go
package types

// SkillPublicationSkill is the reviewed body accepted by the organization
// publication surface. It intentionally mirrors the bounded agent-pack body;
// authority, scope, and hashes are server-derived and never accepted from the
// wire.
type SkillPublicationSkill = AgentConfigAgentPackItem

// SkillPublicationMetadata is the content-free publication projection. No
// skill body or support bytes are present in this type.
type SkillPublicationMetadata struct {
	PublicationID string `json:"publication_id"`
	RevisionID    string `json:"revision_id"`
	Name          string `json:"name"`
	ContentHash   string `json:"content_hash"`
	State         string `json:"state"`
	Generation    uint64 `json:"generation"`
	RuntimeID     string `json:"runtime_id"`
}

// SkillPublicationReference is the exact pinned reference installed on one
// target Agent. It contains no body.
type SkillPublicationReference struct {
	AgentID       string `json:"agent_id"`
	PublicationID string `json:"publication_id"`
	RevisionID    string `json:"revision_id"`
	ContentHash   string `json:"content_hash"`
	Generation    uint64 `json:"generation"`
	RuntimeID     string `json:"runtime_id"`
	State         string `json:"state"`
}

// SkillPublicationReceipt is a content-free mutation receipt safe to replay.
type SkillPublicationReceipt struct {
	OperationID   string `json:"operation_id"`
	Operation     string `json:"operation"`
	PublicationID string `json:"publication_id,omitempty"`
	RevisionID    string `json:"revision_id,omitempty"`
	Generation    uint64 `json:"generation"`
	State         string `json:"state"`
	BeforeHash    string `json:"before_hash,omitempty"`
	AfterHash     string `json:"after_hash,omitempty"`
}

type SkillPublicationPublishRequest struct {
	Identity       IdentityScope         `json:"identity"`
	Name           string                `json:"name"`
	Skill          SkillPublicationSkill `json:"skill"`
	IdempotencyKey string                `json:"idempotency_key"`
	ExpectedAbsent bool                  `json:"expected_absent"`
}

type SkillPublicationPublishResponse struct {
	Publication     SkillPublicationMetadata `json:"publication"`
	Receipt         SkillPublicationReceipt  `json:"receipt"`
	ProtocolVersion string                   `json:"protocol_version"`
}

type SkillPublicationListRequest struct {
	Identity IdentityScope `json:"identity"`
}
type SkillPublicationListResponse struct {
	Publications    []SkillPublicationMetadata `json:"publications,omitempty"`
	ProtocolVersion string                     `json:"protocol_version"`
}

type SkillPublicationGetRequest struct {
	Identity      IdentityScope `json:"identity"`
	PublicationID string        `json:"publication_id"`
}
type SkillPublicationGetResponse struct {
	Publication     SkillPublicationMetadata `json:"publication"`
	ProtocolVersion string                   `json:"protocol_version"`
}

type SkillPublicationSuccessorRequest struct {
	Identity            IdentityScope         `json:"identity"`
	PublicationID       string                `json:"publication_id"`
	ExpectedGeneration  uint64                `json:"expected_generation"`
	ExpectedContentHash string                `json:"expected_content_hash"`
	Skill               SkillPublicationSkill `json:"skill"`
	IdempotencyKey      string                `json:"idempotency_key"`
}
type SkillPublicationSuccessorResponse struct {
	Publication     SkillPublicationMetadata `json:"publication"`
	Receipt         SkillPublicationReceipt  `json:"receipt"`
	ProtocolVersion string                   `json:"protocol_version"`
}

type SkillPublicationRetireRequest struct {
	Identity            IdentityScope `json:"identity"`
	PublicationID       string        `json:"publication_id"`
	ExpectedGeneration  uint64        `json:"expected_generation"`
	ExpectedContentHash string        `json:"expected_content_hash"`
	IdempotencyKey      string        `json:"idempotency_key"`
}
type SkillPublicationRetireResponse struct {
	Publication     SkillPublicationMetadata `json:"publication"`
	Receipt         SkillPublicationReceipt  `json:"receipt"`
	ProtocolVersion string                   `json:"protocol_version"`
}

type SkillPublicationAvailableRequest struct {
	Identity IdentityScope `json:"identity"`
}
type SkillPublicationAvailableResponse struct {
	Publications    []SkillPublicationMetadata `json:"publications,omitempty"`
	ProtocolVersion string                     `json:"protocol_version"`
}

type SkillPublicationInstallRequest struct {
	Identity       IdentityScope `json:"identity"`
	AgentID        string        `json:"agent_id"`
	PublicationID  string        `json:"publication_id"`
	RevisionID     string        `json:"revision_id"`
	ExpectedAbsent bool          `json:"expected_absent"`
	IdempotencyKey string        `json:"idempotency_key"`
}
type SkillPublicationInstallResponse struct {
	Reference       SkillPublicationReference `json:"reference"`
	Receipt         SkillPublicationReceipt   `json:"receipt"`
	ProtocolVersion string                    `json:"protocol_version"`
}

type SkillPublicationUpdateRequest struct {
	Identity            IdentityScope `json:"identity"`
	AgentID             string        `json:"agent_id"`
	PublicationID       string        `json:"publication_id"`
	RevisionID          string        `json:"revision_id"`
	ExpectedGeneration  uint64        `json:"expected_generation"`
	ExpectedContentHash string        `json:"expected_content_hash"`
	IdempotencyKey      string        `json:"idempotency_key"`
}
type SkillPublicationUpdateResponse struct {
	Reference       SkillPublicationReference `json:"reference"`
	Receipt         SkillPublicationReceipt   `json:"receipt"`
	ProtocolVersion string                    `json:"protocol_version"`
}

type SkillPublicationRemoveRequest struct {
	Identity            IdentityScope `json:"identity"`
	AgentID             string        `json:"agent_id"`
	ExpectedGeneration  uint64        `json:"expected_generation"`
	ExpectedContentHash string        `json:"expected_content_hash"`
	IdempotencyKey      string        `json:"idempotency_key"`
}
type SkillPublicationRemoveResponse struct {
	Receipt         SkillPublicationReceipt `json:"receipt"`
	ProtocolVersion string                  `json:"protocol_version"`
}

type SkillPublicationReferencesListRequest struct {
	Identity IdentityScope `json:"identity"`
}
type SkillPublicationReferencesListResponse struct {
	References      []SkillPublicationReference `json:"references,omitempty"`
	ProtocolVersion string                      `json:"protocol_version"`
}

File: web/console/src/lib/protocol/skill-publications.ts
/**
 * Skill-publication wire types — the `skills.publications.*` Protocol family.
 *
 * These interfaces mirror `internal/protocol/types/skill_publications.go`
 * field-for-field. The Go single source and the committed wire manifest are
 * mechanically checked against this hand-maintained Console module by
 * `make protocol-ts-gen-check`.
 *
 * `SkillPublicationSkill` is a Go alias of `AgentConfigAgentPackItem`, so the
 * canonical wire manifest uses that existing interface directly rather than
 * introducing a second TypeScript shape.
 */

import type { AgentConfigAgentPackItem } from './agentconfig.js';
import type { IdentityScope } from './memory-types.js';

export interface SkillPublicationAvailableRequest {
	identity: IdentityScope;
}

export interface SkillPublicationAvailableResponse {
	publications?: SkillPublicationMetadata[];
	protocol_version: string;
}

export interface SkillPublicationGetRequest {
	identity: IdentityScope;
	publication_id: string;
}

export interface SkillPublicationGetResponse {
	publication: SkillPublicationMetadata;
	protocol_version: string;
}

export interface SkillPublicationInstallRequest {
	identity: IdentityScope;
	agent_id: string;
	publication_id: string;
	revision_id: string;
	expected_absent: boolean;
	idempotency_key: string;
}

export interface SkillPublicationInstallResponse {
	reference: SkillPublicationReference;
	receipt: SkillPublicationReceipt;
	protocol_version: string;
}

export interface SkillPublicationListRequest {
	identity: IdentityScope;
}

export interface SkillPublicationListResponse {
	publications?: SkillPublicationMetadata[];
	protocol_version: string;
}

export interface SkillPublicationMetadata {
	publication_id: string;
	revision_id: string;
	name: string;
	content_hash: string;
	state: string;
	generation: number;
	runtime_id: string;
}

export interface SkillPublicationPublishRequest {
	identity: IdentityScope;
	name: string;
	skill: AgentConfigAgentPackItem;
	idempotency_key: string;
	expected_absent: boolean;
}

export interface SkillPublicationPublishResponse {
	publication: SkillPublicationMetadata;
	receipt: SkillPublicationReceipt;
	protocol_version: string;
}

export interface SkillPublicationReceipt {
	operation_id: string;
	operation: string;
	publication_id?: string;
	revision_id?: string;
	generation: number;
	state: string;
	before_hash?: string;
	after_hash?: string;
}

export interface SkillPublicationReference {
	agent_id: string;
	publication_id: string;
	revision_id: string;
	content_hash: string;
	generation: number;
	runtime_id: string;
	state: string;
}

export interface SkillPublicationReferencesListRequest {
	identity: IdentityScope;
}

export interface SkillPublicationReferencesListResponse {
	references?: SkillPublicationReference[];
	protocol_version: string;
}

export interface SkillPublicationRemoveRequest {
	identity: IdentityScope;
	agent_id: string;
	expected_generation: number;
	expected_content_hash: string;
	idempotency_key: string;
}

export interface SkillPublicationRemoveResponse {
	receipt: SkillPublicationReceipt;
	protocol_version: string;
}

export interface SkillPublicationRetireRequest {
	identity: IdentityScope;
	publication_id: string;
	expected_generation: number;
	expected_content_hash: string;
	idempotency_key: string;
}

export interface SkillPublicationRetireResponse {
	publication: SkillPublicationMetadata;
	receipt: SkillPublicationReceipt;
	protocol_version: string;
}

export interface SkillPublicationSuccessorRequest {
	identity: IdentityScope;
	publication_id: string;
	expected_generation: number;
	expected_content_hash: string;
	skill: AgentConfigAgentPackItem;
	idempotency_key: string;
}

export interface SkillPublicationSuccessorResponse {
	publication: SkillPublicationMetadata;
	receipt: SkillPublicationReceipt;
	protocol_version: string;
}

export interface SkillPublicationUpdateRequest {
	identity: IdentityScope;
	agent_id: string;
	publication_id: string;
	revision_id: string;
	expected_generation: number;
	expected_content_hash: string;
	idempotency_key: string;
}

export interface SkillPublicationUpdateResponse {
	reference: SkillPublicationReference;
	receipt: SkillPublicationReceipt;
	protocol_version: string;
}
